parlai/agents/examples/seq2seq.py

```diff
@@ -0,0 +1,202 @@
+    """
+    ExampleModel implements the final components of TorchGeneratorModel.
+
+    Mainly needs to implement reorder_encoder_size, and instantiate self.encoder and
+    self.decoder.
+    """
+
+    def __init__(self, dictionary, esz=256, hidden_size=1024):
+        super().__init__(
+            padding_idx=dictionary[dictionary.null_token],
+            start_idx=dictionary[dictionary.start_token],
+            end_idx=dictionary[dictionary.end_token],
+            unknown_idx=dictionary[dictionary.unk_token],
+        )
+        self.embeddings = nn.Embedding(len(dictionary), esz)
+        self.encoder = Encoder(self.embeddings, hidden_size)
+        self.decoder = Decoder(self.embeddings, hidden_size)
+
+    def output(self, decoder_output):
+        """
+        Perform the final output -> logits transformation.
+        """
+        return F.linear(decoder_output, self.embeddings.weight)
+
+    def reorder_encoder_states(self, encoder_states, indices):
+        """
+        Reorder the encoder states to select only the given batch indices.
+
+        Since encoder_state can be arbitrary, you must implement this yourself.
+        Typically you will just want to index select on the batch dimension.
+        """
+        h, c = encoder_states
+        return h[:, indices, :], c[:, indices, :]
+
+    def reorder_decoder_incremental_state(self, incr_state, indices):
+        """
+        Reorder the decoder states to select only the given batch indices.
+
+        This method can be a stub which always returns None; this will result in the
+        decoder doing a complete forward pass for every single token, making generation
+        O(n^2). However, if any state can be cached, then this method should be
+        implemented to reduce the generation complexity to O(n).
+        """
+        h, c = incr_state
+        return h[:, indices, :], c[:, indices, :]
+
+
+class Seq2seqAgent(tga.TorchGeneratorAgent):
+    """
+    Example agent.
+
+    Implements the interface for TorchGeneratorAgent. The minimum requirement is that it
+    implements ``build_model``, but we will want to include additional command line
+    parameters.
+    """
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        """
+        Add CLI arguments.
+        """
+        # Make sure to add all of TorchGeneratorAgent's arguments
+        super(Seq2seqAgent, cls).add_cmdline_args(argparser)
+
+        # Add custom arguments only for this model.
+        group = argparser.add_argument_group('Example TGA Agent')
+        group.add_argument(
+            '-hid', '--hidden-size', type=int, default=1024, help='Hidden size.'
+        )
+
+    def build_model(self):
+        """
+        Construct the model.
+        """
+
+        model = ExampleModel(self.dict, self.opt['hidden_size'])
+        # we're responsible for setting the embeddings ourselves, but TorchAgent
+        # gives us a nice helper
+        if self.opt['embedding_type'] != 'random':
```

Comment:
why do we need these lines? doesn't TGA handle this for you? CC @stephenroller

Reply:
No, particularly because you could have 3 sets of embeddings: encoder input, decoder input, and decoder output. we typically tie them but the seq2seq model supports not tying them.

I think the `!= random` line can be removed though, or at least moved into `_copy_embeddings`

Reply:
> No, particularly because you could have 3 sets of embeddings: encoder input, decoder input, and decoder output. we typically tie them but the seq2seq model supports not tying them.
> 
> I think the `!= random` line can be removed though, or at least moved into `_copy_embeddings`

Yes, that's cleaner - changing this